lean4-infoview/src/infoview/messages.tsx

```diff
@@ -81,15 +76,36 @@ export function MessagesAtFile({uri, messages}: MessagesAtFileProps) {
     );
 }
 
+function lazy<T>(f: () => T): () => T {
+    let state: {t: T} | undefined
+    return () => {
+        if (!state) state = {t: f()}
+        return state.t
+    }
+}
+
 /** Displays all messages for the specified file. Can be paused. */
 export function AllMessages({uri: uri0}: { uri: DocumentUri }) {
     const ec = React.useContext(EditorContext);
-    const dc = React.useContext(DiagnosticsContext);
+    const sv = React.useContext(VersionContext);
+    const rs = React.useContext(RpcContext);
+    const dc = React.useContext(LspDiagnosticsContext);
     const config = React.useContext(ConfigContext);
-    let diags0 = dc.get(uri0);
-    if (!diags0) diags0 = [];
+    const diags0 = dc.get(uri0) || [];
+
+    let iDiags0 = React.useMemo(() => lazy(async () => {
```

Comment:
I'm not sure this is the right way to do this. Per per React docs: "Remember that the function passed to `useMemo` runs during rendering. Don’t do anything there that you wouldn’t normally do while rendering. For example, side effects belong in `useEffect`, not `useMemo`.".

Reply:
The memo is fine here.  The `lazy` function is side effect-free, it only creates the thunk but does not execute it.

But maybe we should use useEffect for the iDiags() call below..